EIPS/eip-2678.md

```diff
@@ -0,0 +1,2132 @@
+<td><p>Key</p></td>
+<td><p><code>links</code></p></td>
+</tr>
+<tr class="odd">
+<td><p>Type</p></td>
+<td><p>Object (String: String)</p></td>
+</tr>
+</tbody>
+</table>
+
+<div id="sources-object"></div>
+
+The *Sources* Object
+====================
+
+A *Sources* object is defined to have the following fields.
+
+<table>
+<colgroup>
+<col style="width: 50%" />
+<col style="width: 50%" />
+</colgroup>
+<tbody>
+<tr class="odd">
+<td><p>Key</p></td>
+<td><p>A unique identifier for the source file. (String)</p></td>
+</tr>
+<tr class="even">
+<td><p>Value</p></td>
+<td><p><a href="#source-object">Source Object</a></p></td>
+</tr>
+</tbody>
+</table>
+
+
+<div id="source-object"></div>
+
+The *Source* Object
+===================
+
+#### Checksum: `checksum`
+
+Hash of the source file.
+
+<table>
+<colgroup>
+<col style="width: 50%" />
+<col style="width: 50%" />
+</colgroup>
+<tbody>
+<tr class="odd">
+<td><p>Required</p></td>
+<td><p><strong>If</strong> there are no URLs present that contain a content hash.</p></td>
+</tr>
+<tr class="even">
+<td><p>Key</p></td>
+<td><p><code>checksum</code></p></td>
+</tr>
+<tr class="odd">
+<td><p>Value</p></td>
+<td><p><a href="#checksum-object">Checksum Object</a></p></td>
+</tr>
+</tbody>
+</table>
+
+
+#### URLS: `urls`
+
+Array of urls that resolve to the same source file.  
+-   Urls **should** be stored on a content-addressable filesystem.
+    **If** they are not, then either `content` or `checksum` **must** be
+    included.
+
+-   Urls **must** be prefixed with a scheme.
+
+-   If the resulting document is a directory the key **should** be
+    interpreted as a directory path.
+
+-   If the resulting document is a file the key **should** be
+    interpreted as a file path.
+
+<table>
+<colgroup>
+<col style="width: 50%" />
+<col style="width: 50%" />
+</colgroup>
+<tbody>
+<tr class="odd">
+<td><p>Required</p></td>
+<td><p>If <code>content</code> is not included.</p></td>
+</tr>
+<tr class="even">
+<td><p>Key</p></td>
+<td><p><code>urls</code></p></td>
+</tr>
+<tr class="odd">
+<td><p>Value</p></td>
+<td><p>Array(String)</p></td>
+</tr>
+</tbody>
+</table>
+
+
+#### Content: `content`
+
+Inlined contract source. If both `urls` and `content` are provided, the `content` value
+**must** match the content of the files identified in `urls`.
+
+<table>
+<colgroup>
+<col style="width: 50%" />
+<col style="width: 50%" />
+</colgroup>
+<tbody>
+<tr class="odd">
+<td><p>Required</p></td>
+<td><p>If <code>urls</code> is not included.</p></td>
+</tr>
+<tr class="even">
+<td><p>Key</p></td>
+<td><p><code>content</code></p></td>
+</tr>
+<tr class="odd">
+<td><p>Value</p></td>
+<td><p>String</p></td>
+</tr>
+</tbody>
+</table>
+
+
+#### Install Path: `installPath`
+
+Filesystem path of source file.  
+-   **Must** be a relative filesystem path that begins with a `./`.
+
+-   **Must** resolve to a path that is within the current virtual
+    working directory.
+
+-   **Must** be unique across all included sources.
+
```

Comment:
Cannot contain `'../`, so as to avoid accessing files outside of the source folder in improper implementations

Reply:
Yeah this makes sense, though in the spirit of not enforcing unnecessary requirements, is it necessary?

1. Just a couple of lines above we say that every path must resolve to the virtual directory (aka source folder).

2. A path like `contracts/wallet/../safe-math-lib/SafeMathLib.sol` would be dumb but valid.

We could word it like `Cannot start with "../" ..` but I'm not sure it's necessary to explicitly state this restriction given point #1. Thoughts?

Reply:
I think requiring that it "only access inside it's own directory" is important, but you can probably relax not having `../` to be a simpler check if it's all self-containted

Reply:
Just to be clear, imo `**Must** resolve to a path that is within the current virtual working directory.` serves this restriction. Do you interpret this differently?

Reply:
Must have missed that